### PR TITLE
HW6

### PR DESCRIPTION
**Pull request by RafaelKhammadeev**
### Summary

A brief description of the pull request.

### How it works

Screenshots/screencasts of the pull request introduced functionality.

### Test plan

List of steps to manually test introduced functionality:

- Log in
- Go to http://localhost:3000

### Deploy notes

Notes regarding deployment the contained body of work.
These should note any db migrations, services, scripts, etc.

### Checklist:

- [ ] I have performed a self-review of my own code
- [ ] My pull request is shorter than 500 lines

### CODE REVIEW

File: app/policies/task_policy.rb [hunk 1]
```diff
@@ -0,0 +1,31 @@
+class TaskPolicy < ApplicationPolicy
+  authorize :user, allow_nil: true
+
+  def index?
+    show?
```
Comment (RasimKhusaenov) [inline app/policies/task_policy.rb:5]: This is wrong. You can use one method for all, but it shouldn't be method for authorizing action like `show?`

File: app/views/comments/_form.html.erb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+<%= form_with(model: comment, url: [project, task, comment]) do |form| %>
```
Comment (RasimKhusaenov) [inline app/views/comments/_form.html.erb:1]:
```suggestion
<%= form_with(model: [project, task, comment]) do |form| %>
```

File: app/views/comments/_form.html.erb [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  <div class="field">
+    <strong><%= form.label :content, "Content of Comment" %></strong>
+    <%= form.text_area :content %>
+  </div>
+
+  <div class="field">
+    <%= form.label :task_id %>
+    <%= form.select :task_id, Task.all.pluck(:title, :id) %>
```
Comment (RasimKhusaenov) [inline app/views/comments/_form.html.erb:9]: All users can see all tasks in the site?

File: app/views/comments/_form.html.erb [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+  </div>
+
+  <div class="field">
+    <%= form.label :user_id %>
+    <%= form.select :user_id, User.all.pluck(:email, :id) %>
```
Comment (RasimKhusaenov) [inline app/views/comments/_form.html.erb:14]: All users can see another users in the site?

File: app/views/comments/edit.html.erb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+<h1>Editing Project</h1>
+
+<%= render 'form', project: @project, task: @task, comment: @comment %>
```
Comment (RasimKhusaenov) [inline app/views/comments/edit.html.erb:3]: 👍

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -1,21 +1,50 @@
 <!DOCTYPE html>
```
Comment (RasimKhusaenov) [inline app/views/layouts/application.html.erb:1]: This code does not relate to this homework.

File: app/views/tasks/_form.html.erb [hunk 7]
```diff
@@ -26,6 +26,11 @@
     <%= form.datetime_select :deadline_at %>
   </div>
 
+  <div class="field">
+    <%= form.label :status %>
+    <%= form.select :status, ['Not started', 'Started', 'Finished'] %>
```
Comment (RasimKhusaenov) [inline app/views/tasks/_form.html.erb:31]: You should take values from model

File: app/views/tasks/index.html.erb [hunk 8]
```diff
@@ -7,10 +7,11 @@
   <tr>
     <th>Title</th>
     <th>Description</th>
+    <th>Stage</th>
```
Comment (RasimKhusaenov) [inline app/views/tasks/index.html.erb:10]: Stage or status?

File: app/views/tasks/show.html.erb [hunk 9]
```diff
@@ -23,5 +28,28 @@
...
+    <th>Content</th>
+    <th>User</th>
+  </tr>
+  </thead>
+
+  <tbody>
+  <% @task.comments.each do |comment| %>
+    <tr>
+      <td><%= comment.content %></td>
+      <td><%= User.find(comment.user_id)['email'] %></td>
```
Comment (RasimKhusaenov) [inline app/views/tasks/show.html.erb:46]: Does it work?

Comment (RasimKhusaenov) [reply app/views/tasks/show.html.erb:46]: And why not
```suggestion
      <td><%= comment.user.email %></td>
```
?

File: db/migrate/20221030200643_add_status_to_tasks.rb [hunk 10]
```diff
@@ -0,0 +1,5 @@
+class AddStatusToTasks < ActiveRecord::Migration[6.1]
+  def change
+    add_column :tasks, :status, :string, null: false, default: "Not started"
```
Comment (RasimKhusaenov) [inline db/migrate/20221030200643_add_status_to_tasks.rb:3]: Where is it in model?

**Review by RasimKhusaenov — CHANGES_REQUESTED** [review]

File: app/controllers/comments_controller.rb [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  def edit
+  end
+
+  def create
+    @comment = Comment.new(comment_params)
+
+    if @comment.save
+      redirect_to project_task_path(@project, @task), notice: "Comment was successfully created."
+    else
+      @comment.destroy
```
[the remark this answers is not in the document]
Comment (RafaelKhammadeev) [reply app/controllers/comments_controller.rb]: fix it

File: app/controllers/comments_controller.rb [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+      flash.now[:alert] = "Comment must not be empty."
+      render :"tasks/show"
```
[the remark this answers is not in the document]
Comment (RafaelKhammadeev) [reply app/controllers/comments_controller.rb]: fix it

File: app/controllers/comments_controller.rb [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  def create
+    @comment = Comment.new(comment_params)
+
+    if @comment.save
+      redirect_to project_task_path(@project, @task), notice: "Comment was successfully created."
+    else
+      @comment.destroy
+      flash.now[:alert] = "Comment must not be empty."
```
[the remark this answers is not in the document]
Comment (RafaelKhammadeev) [reply app/controllers/comments_controller.rb]: change alert

File: app/views/tasks/index.html.erb (hunk 8 above)
Comment (RafaelKhammadeev) [reply app/views/tasks/index.html.erb]: Status, fix it

File: app/views/layouts/application.html.erb (hunk 6 above)
Comment (RafaelKhammadeev) [reply app/views/layouts/application.html.erb:1]: delete it

File: app/views/tasks/_form.html.erb (hunk 7 above)
Comment (RafaelKhammadeev) [reply app/views/tasks/_form.html.erb]: fix it

File: app/controllers/comments_controller.rb [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+  def set_project
+    @project = Project.find(params[:project_id])
+  end
+
+  def set_comment
+    @comment = Comment.find(params[:id])
+  end
+
+  def comment_params
+    params.require(:comment).permit(:id, :content, :task_id, :user_id)
```
[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/controllers/comments_controller.rb:50]: @RafaelKhammadeev

File: app/controllers/sessions_controller.rb [hunk 15]
```diff
@@ -21,6 +23,12 @@ def create
     end
   end
 
+  def destroy
+    session.delete(:user_id)
+    @current_user = nil
```
[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/controllers/sessions_controller.rb:28]: @RafaelKhammadeev

File: app/controllers/tasks_controller.rb [hunk 16]
```diff
@@ -1,6 +1,7 @@
 class TasksController < ApplicationController
   before_action :set_task, only: %i[show edit update destroy]
-  before_action :set_project, only: %i[new index create destroy]
+  before_action :set_project
+  before_action -> { authorize! @project, with: TaskPolicy }
```
[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/controllers/tasks_controller.rb:4]: @RafaelKhammadeev

File: app/policies/comment_policy.rb [hunk 17]
```diff
@@ -0,0 +1,19 @@
+class CommentPolicy < ApplicationPolicy
+  authorize :user, allow_nil: true
```
[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/policies/comment_policy.rb:2]: @RafaelKhammadeev

File: app/policies/comment_policy.rb [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+
+  def update?
+    true
```
[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/policies/comment_policy.rb:5]: @RafaelKhammadeev

File: app/policies/project_policy.rb [hunk 19]
```diff
@@ -14,6 +14,19 @@ def new?
...
+  def edit?
+    update?
+  end
+
+  def update?
+    record.project_memberships.find_by(user: user).present?
+  end
+
+  def destroy?
+    user.present? && record.project_memberships.find_by(user: user).role == "owner"
```
[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/policies/project_policy.rb:29]: @RafaelKhammadeev

File: app/policies/task_policy.rb [hunk 20]
```diff
@@ -0,0 +1,31 @@
+class TaskPolicy < ApplicationPolicy
+  authorize :user, allow_nil: true
```
[the remark this answers is not in the document]
Comment (RasimKhusaenov) [reply app/policies/task_policy.rb:2]: @RafaelKhammadeev

File: app/policies/task_policy.rb (hunk 1 above)
Comment (RasimKhusaenov) [reply app/policies/task_policy.rb:5]: @RafaelKhammadeev

File: app/views/comments/_form.html.erb (hunk 2 above)
Comment (RasimKhusaenov) [reply app/views/comments/_form.html.erb:1]: @RafaelKhammadeev

File: app/views/comments/_form.html.erb (hunk 3 above)
Comment (RasimKhusaenov) [reply app/views/comments/_form.html.erb:9]: @RafaelKhammadeev

File: app/views/comments/_form.html.erb (hunk 4 above)
Comment (RasimKhusaenov) [reply app/views/comments/_form.html.erb:14]: @RafaelKhammadeev

File: app/views/layouts/application.html.erb (hunk 6 above)
Comment (RasimKhusaenov) [reply app/views/layouts/application.html.erb:1]: No, you didn't

File: db/migrate/20221030200643_add_status_to_tasks.rb (hunk 10 above)
Comment (RasimKhusaenov) [reply db/migrate/20221030200643_add_status_to_tasks.rb:3]: @RafaelKhammadeev

**Review by RasimKhusaenov — CHANGES_REQUESTED** [review]

File: app/policies/project_policy.rb [hunk 21]
```diff
@@ -14,6 +14,18 @@ def new?
   end
 
   def create?
-    user.present?
+    true
```
Comment (RasimKhusaenov) [inline app/policies/project_policy.rb:17]: Why?

File: app/views/layouts/application.html.erb (hunk 6 above)
Comment (RasimKhusaenov) [reply app/views/layouts/application.html.erb:1]: No, you didn't

**Pull request closed without merging**